File: python/source/interactive_books/domain/chunk.py
from dataclasses import dataclass, field
from datetime import datetime

from interactive_books.domain._time import utc_now
from interactive_books.domain.errors import BookError, BookErrorCode


@dataclass(frozen=True)
class Chunk:
    id: str
    book_id: str
    content: str
    start_page: int
    end_page: int
    chunk_index: int
    created_at: datetime = field(default_factory=utc_now)

    def __post_init__(self) -> None:
        if self.start_page < 1:
            raise BookError(
                BookErrorCode.INVALID_STATE,
                f"Chunk start_page must be >= 1, got {self.start_page}",
            )
        if self.end_page < self.start_page:
            raise BookError(
                BookErrorCode.INVALID_STATE,
                f"Chunk end_page ({self.end_page}) must be >= start_page ({self.start_page})",
            )

File: swift/InteractiveBooks/Sources/InteractiveBooksCore/Domain/Entities/Chunk.swift
import Foundation

public struct Chunk: Sendable, Equatable {
    public let id: String
    public let bookId: String
    public let content: String
    public let startPage: Int
    public let endPage: Int
    public let chunkIndex: Int
    public let createdAt: Date

    public init(
        id: String,
        bookId: String,
        content: String,
        startPage: Int,
        endPage: Int,
        chunkIndex: Int,
        createdAt: Date = .now,
    ) throws {
        guard startPage >= 1 else {
            throw BookError.invalidState(
                "Chunk start_page must be >= 1, got \(startPage)",
            )
        }
        guard endPage >= startPage else {
            throw BookError.invalidState(
                "Chunk end_page (\(endPage)) must be >= start_page (\(startPage))",
            )
        }
        self.id = id
        self.bookId = bookId
        self.content = content
        self.startPage = startPage
        self.endPage = endPage
        self.chunkIndex = chunkIndex
        self.createdAt = createdAt
    }
}
